service/matching/matching_engine.go

```diff
@@ -837,6 +837,118 @@ func (e *matchingEngineImpl) listTaskQueuePartitions(request *matchingservice.Li
+	ctx context.Context,
+	req *workflowservice.UpdateWorkerVersioningRulesRequest,
+) (*workflowservice.UpdateWorkerVersioningRulesResponse, error) {
+	ns, err := e.namespaceRegistry.GetNamespace(namespace.Name(req.GetNamespace()))
+	if err != nil {
+		return nil, err
+	}
+
+	taskQueue, err := newTaskQueueID(ns.ID(), req.GetTaskQueue(), enumspb.TASK_QUEUE_TYPE_WORKFLOW)
+	if err != nil {
+		return nil, err
+	}
+
+	tqMgr, err := e.getTaskQueuePartitionManager(ctx, taskQueue, normalStickyInfo, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo carly: do we need to update updateOptions.TaskQueueLimitPerBuildId or updateOptions.KnownVersion like in UpdateWorkerBuildIdCompatibility?
+	updateOptions := UserDataUpdateOptions{}
+
+	err = tqMgr.UpdateUserData(ctx, updateOptions, func(data *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
+		clk := data.GetClock()
+		if clk == nil {
+			tmp := hlc.Zero(e.clusterMeta.GetClusterID())
+			clk = tmp
+		}
+		updatedClock := hlc.Next(clk, e.timeSource)
+		var versioningData *persistencespb.VersioningData
+		switch req.GetOperation().(type) {
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertAssignmentRule:
+			versioningData, err = InsertAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertAssignmentRule(),
+				e.config.VersionAssignmentRuleLimitPerQueue(ns.Name().String()),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceAssignmentRule:
+			versioningData, err = ReplaceAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceAssignmentRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule:
+			versioningData, err = DeleteAssignmentRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteAssignmentRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule:
+			versioningData, err = InsertCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetInsertCompatibleRedirectRule(),
+				e.config.VersionRedirectRuleLimitPerQueue(ns.Name().String()),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule:
+			versioningData, err = ReplaceCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetReplaceCompatibleRedirectRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule:
+			versioningData, err = DeleteCompatibleRedirectRule(
+				updatedClock,
+				data.GetVersioningData(),
+				req.GetDeleteCompatibleRedirectRule(),
+			)
+		case *workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId_:
+			// todo carly: later PR
+		}
+		if err != nil {
+			// operation can't be completed due to limits. no action, do not replicate, report error
+			return nil, false, err
+		}
+		// No need to keep tombstones around for replication
```

Comment:
this comment seems to contradict the condition after it. But the way we handle tombstones is different from the old version, I'm not sure if there is a reason not to cleanup regardless of replication.